Consider adding sugar for familiar loop constructs

The `loop { continue { } }` control flow construct of WGSL isn't present in an other language that I know of and makes writing simple loops more verbose than familiar for-loops. We should consider adding sugar for example:

```
for (<init-statement>; <condition-expr>; <continue-statement>) {
    <body-statements>
}
```

could be desugared to

```
{
    <init-statement>
    loop {
        break if <condition-expr>;
        <body-statements>
        continue {
            <continue-statement>;
        }
    }
}
```

Just catching up here.

I think some considerations have been lost along the way.  Particularly:

* the code to compute the condition and the update clauses could include function calls
* those called functions could have very complex control flow graphs.
* I don't know of a single GPU compiler that does *not* fully inline the entry point and its call tree, bottoming out at compiler builtins.

So for example if you start with this in HLSL/GLSL:
  
      int helper() { ... very complicated code ... }
      int stepper() { ... very complicated code... }
      ...
       for (int a =0; helper(a)< n ; a += stepper() ) {
             body
      }

Then you get arbitrarily complex condition evaluation and aribtrarily complex update code. 

This is not so far-fetched for compute applications over irregular domains.  e.g. work-stealing patterns.  For some reason GCD calculation also comes to mind. Hey, let's throw in bitcoin mining too, why not.


With these patterns in mind:
* a comma operator-like construct; or sequence-of-statements construct is not powerful enough to capture the computation for *either* the condition or the update clauses
* trying to place the update clause into the top, and trying to also have it access values defined in the regular part of the body is overly complex and weird.  I appreciate the effort to put more options on the table, but it goes way beyond traditional languages and breaks all sorts of programming culture and expectations about scoping.

Looking at Robin's taxonomy:

1.   Current WGSL proposal with loop and optional continuing construct handles the complex case.

2.   We can't just use a while loop because it doesn't accommodate inlined arbitrarily complex condition code without also duplicating it or outlining it (inverse of inlining).  (Myles already indicated elsewhere the complexity of out-lining is too high.)  Fundamentally the whlie(cond) is doing two things: (a) computing the condition and (b) choosing whether to escape the loop based on the computed condition result. But (a) might be too complex to fit the while() syntax.

3. is a proposal to add "for" as sugar.

4. is a proposal to add comma-operator or some kind of block syntax.  I address this above, arguing it is inadequate.

@kvark already noted the ability of the code in the continuing block to use a value X defined in the body.  That's *one* of the reasons why we propose putting the continuing block at the end, so that normal scoping rules apply, without the need to artificially hoist variables. Basically what you wrote is what you get, exactly without further complication.  (And devs are asking us to keep our flows as simple as possible.)

There's a caveat to this ability to use the identifier.  For example:

      loop  {
          if (cond) { continue; }
          const step : i32 = ....;
          continuing { 
              // invalid to use "step" here because the continue branch bypasses
              // the declaration of "step".
              i = i + step; 
         }
      }

This kind of error should be familiar to programmers because C compilers catch the following error:

```c
$ cat a.c
int moo(int c) {
  switch(c) {
     int y= 8;
    case 0:
     return y;
  }
  return 0;
}
$ g++ -c a.c 
a.c:4:5: error: cannot jump from switch statement to this case label
    case 0:
    ^
a.c:3:10: note: jump bypasses variable initialization
     int y= 8;
         ^
1 error generated.

``` 

The *other* reason we propose to put the continuing block at the end is that it means when reading the code, the only time you scan _backward_ in the text is for a loop backedge.  This is _easier_ to understand, and more welcoming to new programmers.  A "continue" statement is a _forward_ branch in the CFG, and that matches the top-to-bottom order of the code as proposed for WGSL.
I mentioned this in a meeting a couple of weeks ago, but wanted to capture it in writing here.

--- a/wgsl/index.bs
+++ b/wgsl/index.bs
@@ -1243,7 +1243,9 @@ top of the [[#loop-statement]]. The `for` transforms into loop as:
     # TODO(dsinclair): Wrap in a block when/if we get lexical scoping
     var i : i32 = 0;
     loop {
-      break unless (i < 4);
+      if (!(i < 4)) {
+        break;
+      }
 
       if (a == 0) {
         continue;

--- a/wgsl/index.bs
+++ b/wgsl/index.bs
@@ -496,6 +496,7 @@ Note: literals are parsed greedy. This means that for statements like `a -5`
   <tr><td>`FALLTHROUGH`<td>fallthrough
   <tr><td>`FALSE`<td>false
   <tr><td>`FN`<td>fn
+  <tr><td>`FOR`<td>for
   <tr><td>`FRAGMENT`<td>fragment
   <tr><td>`FUNCTION`<td>function
   <tr><td>`IF`<td>if
@@ -533,7 +534,6 @@ The following is a list of keywords which are reserved for future expansion.
     <td>f16
   <tr>
     <td>f64
-    <td>for
     <td>i8
     <td>i16
     <td>i64
@@ -1031,6 +1031,7 @@ statement
   | if_stmt
   | switch_stmt
   | loop_stmt
+  | for_stmt
   | variable_stmt SEMICOLON
   | break_stmt SEMICOLON
   | continue_stmt SEMICOLON
@@ -1134,6 +1135,7 @@ Optionally, the last statement in the loop body may be a
 
 Note: The loop statement is one of the biggest differences from other shader
 languages.
+
 This design directly expresses loop idioms commonly found in compiled code.
 In particular, placing the loop update statements at the end of the loop body
 allows them to naturally use values defined in the loop body.
@@ -1211,15 +1213,46 @@ allows them to naturally use values defined in the loop body.
 </div>
 * <2> The continue construct is placed at the end of the `loop`
 
-### Other Looping Constructs ### {#other-looping}
+### For statement ### {#for-statement}
 
-Note: This is *proposed* but not in the [SHORTNAME] spec yet
+<pre class='def'>
+for_stmt
+  : FOR PAREN_LEFT for_header PAREN_RIGHT BRACKET_LEFT statements BRACKET_RIGHT
+
+for_header
+  : variable_stmt? SEMICOLON logical_or_expression? SEMICOLON assignment_stmt?
+</pre>
 
 The `for(var i : i32 = 0; i < 4; i = i + 1) {}` statement is syntactic sugar on
-top of the [[#loop-statement]].
+top of the [[#loop-statement]]. The `for` transforms into loop as:
 
-The `while(i < 4) {}` is syntactic sugar on top of the [[#loop-statement]]
-where there is no `continuing` construct.
+<div class='example' heading="For to Loop transformation">
+  <xmp>
+    for(var i : i32 = 0; i < 4; i = i + 1) {
+      if (a == 0) {
+        continue;
+      }
+      a = a + 2;
+    }
+
+    Converts too:
+
+    # TODO(dsinclair): Wrap in a block when/if we get lexical scoping
+    var i : i32 = 0;
+    loop {
+      break if (!(i < 4));
+
+      if (a == 0) {
+        continue;
+      }
+      a = a + 2;
+
+      continuing {
+        i = i + 1;
+      }
+    }
+  </xmp>
+</div>
 
 ## Break statement ## {#break-statement}
 

--- a/wgsl/index.bs
+++ b/wgsl/index.bs
@@ -1032,6 +1032,7 @@ statement
   | switch_stmt
   | loop_stmt
   | for_stmt
+  | func_call_stmt SEMICOLON
   | variable_stmt SEMICOLON
   | break_stmt SEMICOLON
   | continue_stmt SEMICOLON
@@ -1220,7 +1221,9 @@ for_stmt
   : FOR PAREN_LEFT for_header PAREN_RIGHT BRACKET_LEFT statements BRACKET_RIGHT
 
 for_header
-  : variable_stmt? SEMICOLON logical_or_expression? SEMICOLON assignment_stmt?
+  : (variable_stmt | assignment_stmt | func_call_stmt)? SEMICOLON
+     logical_or_expression? SEMICOLON
+     (assignment_stmt | func_call_stmt)?
 </pre>
 
 The `for(var i : i32 = 0; i < 4; i = i + 1) {}` statement is syntactic sugar on
@@ -1253,6 +1256,12 @@ top of the [[#loop-statement]]. The `for` transforms into loop as:
     }
   </xmp>
 </div>
+
+## Function call statemnet ## {#func-call-statement}
+<pre class='def'>
+func_call_stmt
+  : IDENT PAREN_LEFT argument_expression_list* PAREN_RIGHT
+</pre>
 
 ## Break statement ## {#break-statement}
 

--- a/wgsl/index.bs
+++ b/wgsl/index.bs
@@ -1243,7 +1243,7 @@ top of the [[#loop-statement]]. The `for` transforms into loop as:
     # TODO(dsinclair): Wrap in a block when/if we get lexical scoping
     var i : i32 = 0;
     loop {
-      break if (!(i < 4));
+      break unless (i < 4);
 
       if (a == 0) {
         continue;

